### DISCUSSION TITLE
쿠키(Cookie)와 세션(Session)

### DISCUSSION THREAD

**Discussion by yeonkkk** [woowacourse-study/2023-tal-study#67] [기술 면접 질답 모음집]
쿠키(Cookie)와 세션(Session)에 대해 설명해주세요

**Comment by yoondgu** [comment]
HTTP는 무상태성이지만, 쿠키와 세션을 통해 상태가 있는 것처럼 동작할 수 있습니다.

먼저 쿠키에 대해 설명해보겠습니다.
서버에서 set-Cookie 헤더를 사용해 쿠키를 포함시켜 응답을 보내면, 클라이언트 측에서는 이를 쿠키 저장소에 저장합니다.  
클라이언트에서는 Cookie 헤더에 쿠키를 담아서 보낼 수 있습니다.

세션에 대해 설명해보겠습니다.
세션은 쿠키를 기반으로 하지만, 상태를 서버 측에서 관리하는 방식입니다.
일정 시간 동안 같은 클라이언트로부터 들어오는 일련의 요구를 하나의 상태로 보고 이를 유지시키는 기술입니다.
쿠키를 통해 받은 상태를 서버의 저장소에 저장해둡니다.

쿠키는 클라이언트 로컬에 저장되고 요청 탈취의 위험이 있는 것에 비해  
세션 방식은 비교적 보안성을 보장할 수 있습니다.
쿠키를 통해서는 보안 상 위험이 없는 식별값(예를 들면 sessionId)만 보내서 상태를 식별한 뒤, 서버에서 처리해주기 때문입니다.  
또, 브라우저가 종료되어도 상태를 유지할 수 있습니다.  
그럼에도 서버의 자원을 사용해야 하기 때문에, 상황에 따라 적절하게 쿠키를 사용할 수도 있습니다.

**Comment by yeonkkk** [comment]
HTTP의 무상태성으로 인해 쿠키와 세션을 사용합니다.

 쿠키는 클라이언트 측에 작은 텍스트 파일로 저장되며 사용자 식별과 환경 설정과 같은 정보를 저장하며, 클라이언트가 같은 웹 사이트를 방문할 때 상태 정보를 보존합니다. 

반면, 세션은 서버 측에서 사용자 상태를 관리하고 세션 ID를 사용하여 클라이언트를 식별하며, 브라우저를 닫거나 로그아웃할 때 종료되며 안전하고 보안성이 높습니다.

 이로써 무상태성을 극복하여 사용자의 상태 정보를 관리합니다.

**Reply by yeonkkk** [reply]
ref. https://hudi.blog/cookie-and-session/
